core/jvm/src/main/java/scalaz/zio/internal/impls/padding/MutableQueueFieldsPadding.java

```diff
@@ -0,0 +1,129 @@
+package scalaz.zio.internal.impls.padding;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import scalaz.zio.internal.MutableConcurrentQueue;
+
+/*
+ * NOTE: these classes need to be implemented in Java, because:
+ *   1) `head` and `tail` need to be naked protected or public fields
+ *   in order to be accessible by `AtomicLongFieldUpdater`.
+ *   2) there doesnt' seems to be a way to expose naked public or
+ *   protected fields in Scala is it generates accessor methods for
+ *   those.
+ *
+ * The classes below provide padding for contended fields in the
+ * [[MutableConcurrentQueue]] layout, specifically `head` and `tail`.
+ *
+ * The layout of a Java object in memory is dynamic and is under
+ * control of the JVM which can shuffle fields to optimize the
+ * object's footprint. However, when the object is used in a
+ * concurrent setting having all fields densely packed may affect
+ * performance since certain concurrently modified fields can fall on
+ * the same cache-line and be subject to _False Sharing_. In such case
+ * we would like to space out _hot_ fields and thus make access more
+ * cache-friendly.
+ *
+ * On x86 one cache-line is 64KiB and we'd need to space out fields by
+ * at least *2* cache-lines because modern processors do pre-fetching,
+ * i.e. get the requested cache-line and the one after. Thus we need
+ * ~128KiB in-between hot fields.
+ *
+ * One solution could be just adding a bunch of long fields, but if
+ * those are defined within the same class as `head` or `tail`, JVM
+ * will likely shuffle the longs and mess up the padding.
+ *
+ * There is a workaround that relies on the fact the fields of a
+ * super-class go before fields of a sub-class in the object's
+ * layout. So, a properly crafted inheritance chain can guarantee
+ * spacing for the hot fields.
```

Comment:
This is awesome!

Reply:
@iravid thanks! However looks like, I forgot there should be 1 between 0 and 2 😄, so padding wasn't 100% correct. Last commit fixes this.